File: leetcode/622-design-circular-queue/main.go
package main

import (
	"fmt"
)

type MyCircularQueue struct {
	Head  int
	Tail  int
	Size  int
	Queue []int
}

/** Initialize your data structure here. Set the size of the queue to be k. */
func Constructor(k int) MyCircularQueue {
	fixedNums := make([]int, k)
	return MyCircularQueue{-1, -1, k, fixedNums}
}

/** Insert an element into the circular queue. Return true if the operation is successful. */
func (this *MyCircularQueue) EnQueue(value int) bool {
	if this.IsFull() {
		return false
	}
	if this.IsEmpty() {
		this.Head = 0
	}
	this.Tail = (this.Tail + 1) % this.Size
	this.Queue[this.Tail] = value
	return true
}

/** Delete an element from the circular queue. Return true if the operation is successful. */
func (this *MyCircularQueue) DeQueue() bool {
	if this.IsEmpty() {
		return false
	}
	// for DeQueue, we dont need to care about the value on nums[i] after we i++,
	// becos we will override it later

	// it means there is only one value in the queue
	if this.Head == this.Tail {
		this.Head = -1
		this.Tail = -1
		return true
	}
	// move forward our head
	this.Head = (this.Head + 1) % this.Size
	return true
}

/** Get the front item from the queue. */
func (this *MyCircularQueue) Front() int {
	if this.IsEmpty() {
		return -1
	}
	return this.Queue[this.Head]
}

/** Get the last item from the queue. */
func (this *MyCircularQueue) Rear() int {
	if this.IsEmpty() {
		return -1
	}
	return this.Queue[this.Tail]
}

/** Checks whether the circular queue is empty or not. */
func (this *MyCircularQueue) IsEmpty() bool {
	return this.Head == -1
}

/** Checks whether the circular queue is full or not. */
func (this *MyCircularQueue) IsFull() bool {
	return (this.Tail+1)%this.Size == this.Head
}

func main() {
	obj := Constructor(3)
	obj.EnQueue(1)
	obj.EnQueue(2)
	fmt.Println(obj.EnQueue(3))
	fmt.Println(obj.EnQueue(4), obj)
	fmt.Println(obj.IsFull())
	fmt.Println(obj.DeQueue())
	fmt.Println(obj.IsFull())
	fmt.Println(obj.EnQueue(4))
	fmt.Println(obj.IsFull())
	fmt.Println(obj.Front())
	fmt.Println(obj.Rear())
}

File: leetcode/622-design-circular-queue/main.py
"""
    1st approach: use an array
    88 ms, faster than 22.55%
"""


class MyCircularQueue(object):

    def __init__(self, k):
        """
        Initialize your data structure here. Set the size of the queue to be k.
        :type k: int
        """
        self.nums = []
        self.cap = k

    def enQueue(self, value):
        """
        Insert an element into the circular queue. Return true if the operation is successful.
        :type value: int
        :rtype: bool
        """
        if self.isFull():
            return False
        self.nums.append(value)
        return True

    def deQueue(self):
        """
        Delete an element from the circular queue. Return true if the operation is successful.
        :rtype: bool
        """
        if self.isEmpty():
            return False
        self.nums.pop(0)
        return True

    def Front(self):
        """
        Get the front item from the queue.
        :rtype: int
        """
        if self.isEmpty():
            return -1
        return self.nums[0]

    def Rear(self):
        """
        Get the last item from the queue.
        :rtype: int
        """
        if self.isEmpty():
            return -1
        return self.nums[-1]

    def isEmpty(self):
        """
        Checks whether the circular queue is empty or not.
        :rtype: bool
        """
        return len(self.nums) == 0

    def isFull(self):
        """
        Checks whether the circular queue is full or not.
        :rtype: bool
        """
        return len(self.nums) == self.cap


"""
    2nd approach: doubly linked list
    168 ms, faster than 5.51%
"""


class DLLNode(object):
    def __init__(self, val):
        self.val = val
        self.next = None
        self.prev = None


class MyCircularQueue(object):

    def __init__(self, k):
        """
        Initialize your data structure here. Set the size of the queue to be k.
        :type k: int
        """
        self.head = DLLNode(0)
        self.tail = DLLNode(0)
        self.head.next = self.tail
        self.tail.prev = self.head
        self.cap = k
        self.count = 0

    def enQueue(self, value):
        """
        Insert an element into the circular queue. Return true if the operation is successful.
        :type value: int
        :rtype: bool
        """
        if self.isFull():
            return False

        last = self.tail.prev
        node = DLLNode(value)
        last.next = node
        node.prev = last
        node.next = self.tail
        self.tail.prev = node
        self.count += 1

        return True

    def deQueue(self):
        """
        Delete an element from the circular queue. Return true if the operation is successful.
        :rtype: bool
        """
        if self.isEmpty():
            return False

        first = self.head.next
        self.head.next = first.next
        first.next.prev = self.head

        self.count -= 1
        return True

    def Front(self):
        """
        Get the front item from the queue.
        :rtype: int
        """
        if self.isEmpty():
            return -1
        return self.head.next.val

    def Rear(self):
        """
        Get the last item from the queue.
        :rtype: int
        """
        if self.isEmpty():
            return -1
        return self.tail.prev.val

    def isEmpty(self):
        """
        Checks whether the circular queue is empty or not.
        :rtype: bool
        """
        return self.count == 0

    def isFull(self):
        """
        Checks whether the circular queue is full or not.
        :rtype: bool
        """
        return self.count == self.cap


"""
    3rd approach: create a fixed-sized queue
    - 2 pointers

    Time    O(1)
    Space   O(1)
    84 ms, faster than 25.83%
"""


class MyCircularQueue(object):

    def __init__(self, k):
        """
        Initialize your data structure here. Set the size of the queue to be k.
        :type k: int
        """
        self.nums = k * [-1]
        self.start = -1
        self.end = -1

    def enQueue(self, value):
        """
        Insert an element into the circular queue. Return true if the operation is successful.
        :type value: int
        :rtype: bool
        """
        if self.isFull():
            return False
        if self.start == -1:
            self.start = 0
        idx = self._calNext(self.end)
        self.nums[idx] = value
        self.end = idx
        return True

    def deQueue(self):
        """
        Delete an element from the circular queue. Return true if the operation is successful.
        :rtype: bool
        """
        if self.isEmpty():
            return False
        self.nums[self.start] = -1
        idx = self._calNext(self.start)
        temp = self.nums[idx]
        if self.start == self.end:
            self.start = -1
            self.end = -1
        else:
            self.start = idx
        # return temp # if question asks to return the val
        return True

    def Front(self):
        """
        Get the front item from the queue.
        :rtype: int
        """
        if self.isEmpty():
            return -1
        return self.nums[self.start]

    def Rear(self):
        """
        Get the last item from the queue.
        :rtype: int
        """
        if self.isEmpty():
            return -1
        return self.nums[self.end]

    def isEmpty(self):
        """
        Checks whether the circular queue is empty or not.
        :rtype: bool
        """
        return self.start == -1

    def isFull(self):
        """
        Checks whether the circular queue is full or not.
        :rtype: bool
        """
        return self.start == self._calNext(self.end)

    def _calNext(self, n):
        k = len(self.nums)
        return (n + 1) % k
